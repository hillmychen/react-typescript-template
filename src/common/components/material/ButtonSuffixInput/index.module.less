.captchaButtonContainer {
  position: relative;
  display: inline-block;
  width: 100%;

  .captchaButton {
    height: initial;
    position: absolute;
    right: 4px;
    top: 4px;
    bottom: 4px;
  }
}
