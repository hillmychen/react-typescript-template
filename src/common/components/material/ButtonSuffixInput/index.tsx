import React from 'react'
import { Input } from '../../basic/Input'
import { AutoLoadingButton } from '../../material/AutoLoadingButton'
import styles from './index.module.less'

export interface ButtonSuffixInputProps {
  value?: string
  onChange?: (value: React.ChangeEvent<HTMLInputElement>) => void
  disabled?: boolean
  maxLength?: number
  trim?: boolean
  buttonText?: string
  buttonDisabled?: boolean
  buttonLoading?: boolean
  placeholder?: string
  onClick?: (value: string | undefined) => void | Promise<void>
}

export class ButtonSuffixInput extends React.PureComponent<ButtonSuffixInputProps> {
  public static defaultProps: PickOptional<ButtonSuffixInputProps> = {
    buttonText: intl.get('button').d('按钮'),
  }

  onClickButton = () => this.props.onClick?.(this.props.value)

  render() {
    const { value, onChange, disabled, buttonText, buttonDisabled, buttonLoading, placeholder, trim, maxLength } = this.props
    return (
      <div className={styles.captchaButtonContainer}>
        <Input trim={trim} maxLength={maxLength} value={value} onChange={onChange} disabled={disabled} placeholder={placeholder} />
        <AutoLoadingButton type="primary" ghost size="small" className={styles.captchaButton} disabled={buttonDisabled} loading={buttonLoading} onClick={this.onClickButton}>
          {buttonText}
        </AutoLoadingButton>
      </div>
    )
  }
}

export default ButtonSuffixInput
