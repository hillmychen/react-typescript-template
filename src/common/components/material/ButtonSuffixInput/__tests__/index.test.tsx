/*
 * @Author: Hughie
 * @Date: 2020-05-16 18:36:15
 * @LastEditTime: 2020-06-23 15:42:30
 * @LastEditors: Hughie
 * @Description:
 */
import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import ButtonSuffixInput from '../index'
import UserEvent from '@testing-library/user-event'

describe('ButtonSuffixInput', () => {
  test('should support maxLength', async () => {
    const { asFragment } = render(<ButtonSuffixInput maxLength={3} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('buttonDisabled', async () => {
    const { asFragment } = render(<ButtonSuffixInput buttonDisabled />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('trim', async () => {
    const onChange = jest.fn()
    const { asFragment, container } = render(<ButtonSuffixInput onChange={onChange} />)
    // 这里尝试不使用data-testid 来获取 控件
    const firstInput = container.querySelector('input')!
    UserEvent.type(firstInput, '你好')
    expect(onChange).toHaveBeenCalledTimes(2)

    expect(asFragment()).toMatchSnapshot()
  })

  test('onClick', async () => {
    const onClick = jest.fn()
    const { asFragment, container, rerender } = render(<ButtonSuffixInput buttonDisabled onClick={onClick} />)
    expect(asFragment()).toMatchSnapshot()
    const firstButton = container.querySelector('button')!
    UserEvent.click(firstButton)
    expect(onClick).not.toHaveBeenCalled()
    rerender(<ButtonSuffixInput onClick={onClick} />)
    UserEvent.click(firstButton)
    expect(onClick).toHaveBeenCalled()
  })
})
