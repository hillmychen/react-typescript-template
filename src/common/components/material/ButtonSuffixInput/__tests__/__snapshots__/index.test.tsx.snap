// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonSuffixInput buttonDisabled 1`] = `
<DocumentFragment>
  <div
    class="captchaButtonContainer"
  >
    <input
      class="ant-input"
      type="text"
      value=""
    />
    <button
      class="ant-btn ant-btn-primary ant-btn-sm ant-btn-background-ghost captchaButton"
      disabled=""
      type="button"
    >
      <span>
        按 钮
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`ButtonSuffixInput onClick 1`] = `
<DocumentFragment>
  <div
    class="captchaButtonContainer"
  >
    <input
      class="ant-input"
      type="text"
      value=""
    />
    <button
      class="ant-btn ant-btn-primary ant-btn-sm ant-btn-background-ghost captchaButton"
      disabled=""
      type="button"
    >
      <span>
        按 钮
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`ButtonSuffixInput should support maxLength 1`] = `
<DocumentFragment>
  <div
    class="captchaButtonContainer"
  >
    <input
      class="ant-input"
      maxlength="3"
      type="text"
      value=""
    />
    <button
      class="ant-btn ant-btn-primary ant-btn-sm ant-btn-background-ghost captchaButton"
      type="button"
    >
      <span>
        按 钮
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`ButtonSuffixInput trim 1`] = `
<DocumentFragment>
  <div
    class="captchaButtonContainer"
  >
    <input
      class="ant-input"
      type="text"
      value="你好"
    />
    <button
      class="ant-btn ant-btn-primary ant-btn-sm ant-btn-background-ghost captchaButton"
      type="button"
    >
      <span>
        按 钮
      </span>
    </button>
  </div>
</DocumentFragment>
`;
